arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java

```diff
@@ -0,0 +1,325 @@
+
+    private final List<FileScanTask> fileTasks;
+    private final Iterator<FileScanTask> fileItr;
+    private final Map<String, InputFile> inputFiles;
+    private final Schema expectedSchema;
+    private final String nameMapping;
+    private final boolean caseSensitive;
+    private final int batchSize;
+    private final boolean reuseContainers;
+    private CloseableIterator<ColumnarBatch> currentIterator;
+    private ColumnarBatch current;
+    private FileScanTask currentTask;
+
+    /**
+     * Create a new instance.
+     *
+     * @param tasks             Combined file scan tasks.
+     * @param expectedSchema    Read schema. The returned data will have this schema.
+     * @param nameMapping       Mapping from external schema names to Iceberg type IDs.
+     * @param io                File I/O.
+     * @param encryptionManager Encryption manager.
+     * @param caseSensitive     If {@code true}, column names are case sensitive.
+     *                          If {@code false}, column names are not case sensitive.
+     * @param batchSize         Batch size in number of rows. Each Arrow batch contains
+     *                          a maximum of {@code batchSize} rows.
+     * @param reuseContainers   If set to {@code false}, every {@link Iterator#next()} call creates
+     *                          new instances of Arrow vectors.
+     *                          If set to {@code true}, the Arrow vectors in the previous
+     *                          {@link Iterator#next()} may be reused for the data returned
+     *                          in the current {@link Iterator#next()}.
+     *                          This option avoids allocating memory again and again.
+     *                          Irrespective of the value of {@code reuseContainers}, the Arrow vectors
+     *                          in the previous {@link Iterator#next()} call are closed before creating
+     *                          new instances if the current {@link Iterator#next()}.
+     */
+    VectorizedCombinedScanIterator(
+        CloseableIterable<CombinedScanTask> tasks,
+        Schema expectedSchema,
+        String nameMapping,
+        FileIO io,
+        EncryptionManager encryptionManager,
+        boolean caseSensitive,
+        int batchSize,
+        boolean reuseContainers) {
+      this.fileTasks = StreamSupport.stream(tasks.spliterator(), false)
+          .map(CombinedScanTask::files)
+          .flatMap(Collection::stream)
+          .collect(Collectors.toList());
+      this.fileItr = fileTasks.iterator();
+      this.inputFiles = Collections.unmodifiableMap(fileTasks.stream()
+          .flatMap(fileScanTask -> Stream.concat(Stream.of(fileScanTask.file()), fileScanTask.deletes().stream()))
+          .map(file -> EncryptedFiles.encryptedInput(io.newInputFile(file.path().toString()), file.keyMetadata()))
+          .map(encryptionManager::decrypt)
```

Comment:
I think this should use the same logic that's in the [Spark and Flink read paths](https://github.com/apache/iceberg/blob/e8b3d1e8fb50df7ed2662a31d7a212fae87ad2a5/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java#L68-L72). That uses the bulk decryption interface to avoid extra calls to a key manager and avoids problems with duplicate file entries that have been bugs in the past.

Reply:
Thanks for looking into this pr @rdblue.
I'll look into the suggestion and get back to you. A new pr will be cleaner for the suggested changes. Is that fine?

Reply:
Yes, a follow-up is fine with me.

Reply:
This was already fixed in #2720